### ISSUE TITLE
Add opus support

### ISSUE THREAD

**Issue by jplatte** [squentin/gmusicbrowser#77]
It would be nice to be able to play [opus](http://www.opus-codec.org/) files. This should definitely be possible with #76 done, but may also be possible without it.

**Comment by GenesisB** [comment]
I'd like to see this too
--------

**Comment by cbiere** [comment]
I second this. Since there's been support for Opus in gstreamer and mplayer for years, I was quite surprised when I found out, that all my Opus files are being ignored.

**Comment by nicolas-raoul** [comment]
Apparently GStreamer support has been added about 2 years ago (#76).

But unfortunately gmusicbrowser still does not play Opus files (gmusicbrowser v1.1.14 on Ubuntu 2015.10)

Opus is the format used by `youtube-dl -x` (audio download from YouTube)

**Comment by berarma** [comment]
+1

**Comment by RomanHargrave** [comment]
mpv supports opus, so it's just a matter of modifying GMusicBrowser to include the files in the index.

**Comment by cbiere** [comment]
Switched to Clementine Music Player because of this.

**Comment by RomanHargrave** [comment]
Making some progress hacking on oggheader ![Tantalizing Progress Image](http://i.imgur.com/d8T2zdl.png)

**Comment by davidfoerster** [comment]
@RomanHargrave Opus supports 44.1 kHz sampling rate? Which encoder did you use?

**Comment by RomanHargrave** [comment]
It's the input encoding rate. The ID packet encodes data about what the encoder was fed. As for output rate I think opus is VBR.

**Comment by davidfoerster** [comment]
@RomanHargrave Are we talking about the same thing? VBR is "variable bit rate" which is entirely different from the sampling rate.

Which audio (decoder) framework do you intend to target? Or are you only going to work on file headers?

**Comment by RomanHargrave** [comment]
I wasn't thinking. I edited my comments before you, heh. 

Due to the way GMusicBrowser handles files, you only need to implement a tag/metadata parser for each format. The file itself is handed as-as to the decoder (mpv, gstreamer, etc). mpv and gstreamer should both work, since I know for fact they support opus.

**Comment by RomanHargrave** [comment]
One thing to note is that while it may seem that the length calculation brought over from ogg works, it doesn't quite - it's usually off by ten to twenty seconds.

**Comment by RomanHargrave** [comment]
Tags are read correctly now, I haven't touched writing yet.

**Comment by davidfoerster** [comment]
@RomanHargrave I'm surprised you can't reuse the code that parses and writes the Ogg container data and tags of Vorbis files.

**Comment by RomanHargrave** [comment]
I suppose I could. This was more the result of tinkering with things.

**Comment by RomanHargrave** [comment]
@davidfoerster I've merged the common code.

**Comment by davidfoerster** [comment]
Is there any chance to add better replaygain support? At the moment the GStreamer decoder applies replaygain from the metadata (as it should) but GMB is unaware of it and thus applies the configured fall-back gain regardless.

I'm attaching a [sample file](https://github.com/squentin/gmusicbrowser/files/1903505/103-Saturnine.zip) in Opus and Vorbis  format (both from the CC-licensed [*Machina II* by the Smashing Pumpkins](https://archive.org/details/machina.II.ver.SPIFC).

**Comment by RomanHargrave** [comment]
@davidfoerster the issue is with missing RG support in non-gstreamer backends? or the issue is with improper RG support in the Opus tag parser?

**Comment by davidfoerster** [comment]
@RomanHargrave I'd say the issue is in the tag parser, not in GStreamer. The latter appears to do what it's supposed to (apply replay gain from file metadata) while GMB applies the fall-back gain for files *without* replay gain info *on top* the replay gain that GStreamer applies.

**Comment by RomanHargrave** [comment]
i'll need to explore the RG implementation in GMB core for that. my backend is MPV (I use jack and have a large collection of different file types, GStreamer is simply out of the question for me). I'll have to investigate a few things:

1. Does RG even work correctly for MPV - if not, fix it

2. Where/why RG problems happen with Opus

**Comment by davidfoerster** [comment]
I'd investigate the MPV back-end too if I knew which Perl bindings or other (Ubuntu) packages I need to install to enable the MPV option in the audio back-end settings of GMB. Any suggestions?

**Comment by RomanHargrave** [comment]
The MPV backend appears to have working RG.
You do not need any special modules installed - all RPCs are done internally. The backend itself uses `JSON::PP`, `Time::HiRes` and `IO::Socket::Unix` which are all most certainly going to be present in the first place.

**Comment by davidfoerster** [comment]
I lacked the `JSON:PP` module and installed it now. The others are part of the Perl base installation. The MPV option in GMB is still unselectable.

**Comment by RomanHargrave** [comment]
```perl
sub check_version
{	return unless $mpv;
	my $output= qx/$mpv -V/;
	my ($ok,$version);
	if ($output=~m/mpv\s*(\d+)\.(\d+)(\S+)?/i) { $ok= $1>0 || $2>=7; $version= "$1.$2".($3||"") } #requires version 0.7 or later
	elsif ($output=~m/mpv\s*(git-[[:xdigit:]]+)/i) { $ok=1; $version=$1; } #assume git version is ok
	else { warn "mpv: error looking up mpv version\n"; }
	warn "mpv: found mpv version $version\n" if $version && ($::debug || !$ok);
	if (!$ok) { warn "mpv version earlier than 0.7 are not supported -> mpv backend disabled\n" }
	return $ok;
}
```

do you have ≥0.7?

**Comment by RomanHargrave** [comment]
@davidfoerster I've just performed EBU loudness measurements of the same track played in MPV by GMB, first in ogg, then opus.

RG mode is set to album, and the two files have identical RG tags.

OGG Measurement:

![Ogg](https://i.imgur.com/2B84ed6.png)

Spectrum w/ Peaks:

![OGG Spectrum](https://i.imgur.com/PnJ1mOl.png)

[Screencap of OGG file properties](https://i.imgur.com/cuhagHk.png)

Opus Measurement:

![Opus](https://i.imgur.com/Tz3ndMA.png)

Spectrum w/ Peaks:

![Opus Spectrum](https://i.imgur.com/dnCxcxr.png)

[Screencap of Opus file properties](https://i.imgur.com/6gcfh9g.png)

Comparing the two spectrums, it would appear that any visible difference in volume between these during playback is within an error margin.

![Deviation](https://i.imgur.com/lo6Rptn.png)

Capture Setup:

![Capture Setup](https://i.imgur.com/KNLmqFO.png)

**Comment by RomanHargrave** [comment]
FBG is 0.0dB but swapping between album/track mode causes noticeable changes for both formats.

MPV invocation is

`/usr/bin/mpv --input-unix-socket=/home/roman/.config/gmusicbrowser/gmb_mpv_sock --idle --no-video --no-input-terminal --really-quiet --gapless-audio=weak --softvol-max=100 --mute=no --no-sub-auto --volume=55 --af-add=@RG:volume=0:replaygain-album=yes:replaygain-clip=yes:replaygain-preamp=1 --replaygain album --af=channels=2 --ao jack --jack-name GMusicBrowser`

I'll have to start it manually and see if can catch GMB changing to fallback gain, though I doubt it.

**Comment by RomanHargrave** [comment]
On a more pressing note, I've noticed that GMB is writing bad tag headers for certain opus files. This certainly warrants a more critical investigation than potential RG issues.

**Comment by squentin** [comment]
Better late than never I guess.
It's fixed in both branches: master (gtk3) and v1.1.x (gtk2)
Please report any problems you encounter with opus files.
